### ISSUE TITLE
Improve  XML External Entities (XXE) detection

### ISSUE THREAD

**Issue by studersi** [coreruleset/coreruleset#2502]
### Motivation

I could be wrong but as far as I can tell, the XML External Entity (XXE) protection offered by CRS is fairly poor. There have been multiple attempts in the past to address these shortcomings but they have not succeeded and there are currently no ongoing efforts to improve the situation.

These are the issues and pull requests that have been created in the past:
* https://github.com/coreruleset/coreruleset/issues/1319
* https://github.com/coreruleset/coreruleset/issues/1320
* https://github.com/coreruleset/coreruleset/pull/2061
* https://github.com/coreruleset/coreruleset/pull/2163

From the discussions in the tickets and pull requests above, the problem appears to be that there is a shortcoming of ModSecurtiy which causes the targeted collections (`REQUEST_BODY` and `FULL_REQUEST`) to be empty if a body parser has been activated. This, of course, would be the case in most situation where XXE detection is desired, braking the proposed rules.

However, since the information related to XXEs is contained in `<!DOCTYPE ...>` tags, which are not accessible through the XML collection using XPath expressions, is not possible to write reliable ModSecurity rules to detect XXEs for these requests.

This issue is known but considering the following discussion, it is unlikely to be fixed in ModSecurity any time soon: https://github.com/SpiderLabs/ModSecurity/issues/2087.

There have been comments that a plugin leveraging ModSecurity's Lua capabilities might help but as far as I can tell, this was never implemented: https://github.com/coreruleset/coreruleset/pull/2163#issuecomment-997432471.

### Proposed solution

Implement a plugin similar to the [antivirus plugin](https://github.com/coreruleset/antivirus-plugin) which can reliably detect XXEs in requests independently of whether or not a body parser has been activated.

### Alternatives

A cleaner approach might be to create a pull request to ModSecurity to provide additional information about the payload like the DOCTYPE through either the existing XML collection or through a new collection. This would allow for regular rules to be written in order to detect XXEs. However, this would require a change to ModSecurity, which is outside the control of the Core Rule Set Project.

### Additional context

There is a ModSecurity directive related to XXEs: [SecXmlExternalEntity](https://github.com/SpiderLabs/ModSecurity/wiki/Reference-Manual-(v2.x)#secxmlexternalentity). This directive is turned off by default. From what I understand, this directive only configures whether ModSec expands External Entities in XML or not. It does not configure whether XXEs are allowed in general or not. XML containing XXEs are not blocked, even if SecXmlExternalEntity is turned off.

There are rules that might detect XXEs in certain circumstances, however, this is mostly incidental as they do not explicitly check for XXEs.

**Comment by theMiddleBlue** [comment]
+1

**Comment by dune73** [comment]
This is a very good roundup and I like the approach.

@airween : Could you chime in from a ModSec v2 / v3 perspective, namely with regards to `REQUEST_BODY` availability?

**Comment by studersi** [comment]
From what I saw, there is already a discussion going on with respects to differences between v2 and v3: https://github.com/SpiderLabs/ModSecurity/issues/2146.

It appears that there is a bug in v3 where the `REQUEST_BODY` variable is actually set despite having a body processor active. Though there also seem to be discrepancies between the documentation and the v2 implementation, which makes it difficult to tell what the correct behaviour should be.

**Comment by dune73** [comment]
Exactly. It's one of the open quarrels we have with the ModSec devs.

Airween might be able to give us the latest news / perspectives here.

**Comment by airween** [comment]
> @airween : Could you chime in from a ModSec v2 / v3 perspective, namely with regards to `REQUEST_BODY` availability?

I think I can't add anything what @studersi described above perfectly.

Though only one thing: we always think in ModSecurity v2 and v3, but there is a new kid on the ground: Coraza. @jptosso and me already discussed about it, Coraza follows v2's behavior, and `REQUEST_BODY` isn't available if CT is `application/xml` for eg.

Accordingly, the best solution would be use the Lua plugin as I noted in my [mentioned comment](https://github.com/coreruleset/coreruleset/pull/2163#issuecomment-997432471), but I don't know what is the Lua support in Coraza.

**Comment by azurit** [comment]
> but I don't know what is the Lua support in Coraza.

Unfortunately, none. :)

**Comment by airween** [comment]
> It appears that there is a bug in v3

well, I'm not sure it's a bug from point of view of developers, rather it's a feature...

> Though there also seem to be discrepancies between the documentation and the v2 implementation, which makes it difficult to tell what the correct behavior should be.

Yes, you see that perfectly well. One of the options is "fix the *bug*" (if there is), the other one is redefine the behavior. :)

**Comment by dune73** [comment]
@jptosso: What is the prospect of Coraza getting LUA Support on a horizon of 1-2 years?

**Comment by azurit** [comment]
I'm very close to a working solution (which uses Lua) but it's going to be a little more compex. Not sure if it's worth it.

**Comment by dune73** [comment]
@azurit : You meant the LUA solution you have get going might be so complicated that it is not worth to use in production?

**Comment by azurit** [comment]
Not exactly, it will probably be too complex and take much time to implement it correctly. I don't have a working prototype yet but i was able to resolve all hardest problems. If no other problems arises, plugin will work and do this:
 - disabling embeded modsec XML parser with `SecRuleRemoveById 200000`
 - parsing XML from `REQUEST_BODY` at the beginning of phase 2 using Lua script and some of the many XML lua libraries
 - populating `tx.xml.*` so it will look the same as modsec `XML` collection
 - replacing target `XML:/*` with `TX:/^xml\./` in all rules (using `SecRuleUpdateTargetById`)

The hardest step will be populating `tx.xml.*` so the result will be the same as if modsec did it (i.e. we need to simulate modsec in this).

What do you think?

**Comment by jptosso** [comment]
> @jptosso: What is the prospect of Coraza getting LUA Support on a horizon of 1-2 years?

Coraza is taking a different path, we are looking for WASM or something that is more flexible. Still, Lua support can be implemented as a coraza plugin, I think it could take a month if there is a real need. 

One of the main issues is that we haven't found decent documentation on the lua-modsecurity functions and variables.

**Comment by azurit** [comment]
Good news everyone!

Attaching first, buggy, incomplete BUT working version of XXE plugin:
[xxe.tar.gz](https://github.com/coreruleset/coreruleset/files/8533545/xxe.tar.gz)


```
$ cat xxe2.xml
<!--?xml version="1.0" ?-->
<!DOCTYPE replace [<!ENTITY ent SYSTEM "file:///etc/shadow"> ]>
<userInfo>
 <firstName>John</firstName>
 <lastName>union select</lastName>
</userInfo>

$ curl http://example.com/ -H "Content-Type: application/xml" --data-binary "@xxe2.xml" -X POST

[Thu Apr 21 19:43:08.324606 2022] [:error] [pid 2528] [client ...:39096] [client ...] ModSecurity: Warning. Pattern match "<!ENTITY\\\\s+[^\\\\s]+\\\\s+(?:SYSTEM|PUBLIC)\\\\s+['\\"](?i:file|http|https|ftp|php|zlib|data|glob|phar|ssh2|rar|ogg|expect|zip|jdbc)://" at REQUEST_BODY. [file "/usr/share/modsecurity-crs-new/plugins/xxe-before.conf"] [line "19"] [id "20"] [msg "XML eXternal Entity: LFI/RFI using wrapper"] [data "Matched Data: <!--?xml version=\\x221.0\\x22 ?--> <!DOCTYPE replace [<!ENTITY ent SYSTEM \\x22file:///etc/shadow\\x22> ]> <userInfo> <firstName>John</firstName> <lastName>union select</lastName> </userInfo>  found within REQUEST_BODY"] [severity "CRITICAL"] [hostname "example.com"] [uri "/"] [unique_id "YmGXrLPsXuWR3RRLEBQ-0gAAAAE"]

[Thu Apr 21 19:43:08.325087 2022] [:error] [pid 2528] [client ...:39096] [client ...] ModSecurity: Warning. Pattern match "<!ENTITY\\\\s+[^\\\\s]+\\\\s+(?:SYSTEM|PUBLIC)\\\\s+['\\"][^:]+" at REQUEST_BODY. [file "/usr/share/modsecurity-crs-new/plugins/xxe-before.conf"] [line "30"] [id "21"] [msg "XML eXternal Entity: Local File Inclusion attempt"] [data "Matched Data: <!--?xml version=\\x221.0\\x22 ?--> <!DOCTYPE replace [<!ENTITY ent SYSTEM \\x22file:///etc/shadow\\x22> ]> <userInfo> <firstName>John</firstName> <lastName>union select</lastName> </userInfo>  found within REQUEST_BODY"] [severity "CRITICAL"] [hostname "example.com"] [uri "/"] [unique_id "YmGXrLPsXuWR3RRLEBQ-0gAAAAE"]

[Thu Apr 21 19:43:08.328423 2022] [:error] [pid 2528] [client ...:39096] [client ...] ModSecurity: Warning. Pattern match "(?i:(?:[\\"'`](?:;?\\\\s*?(?:having|select|union)\\\\b\\\\s*?[^\\\\s]|\\\\s*?!\\\\s*?[\\"'`\\\\w])|(?:c(?:onnection_id|urrent_user)|database)\\\\s*?\\\\([^\\\\)]*?|u(?:nion(?:[\\\\w(\\\\s]*?select| select @)|ser\\\\s*?\\\\([^\\\\)]*?)|s(?:chema\\\\s*?\\\\([^\\\\)]*?|elect.*?\\\\w?user\\\\()|in ..." at TX:xml.userInfo.lastName. [file "/usr/share/modsecurity-crs-new/rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf"] [line "183"] [id "942190"] [msg "Detects MSSQL code execution and information gathering attempts"] [data "Matched Data: union select found within TX:xml.userInfo.lastName: union select"] [severity "CRITICAL"] [ver "OWASP_CRS/3.3.2"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-sqli"] [tag "paranoia-level/1"] [tag "OWASP_CRS"] [tag "capec/1000/152/248/66"] [tag "PCI/6.5.2"] [hostname "example.com"] [uri "/"] [unique_id "YmGXrLPsXuWR3RRLEBQ-0gAAAAE"]
```

**Comment by studersi** [comment]
@azurit Very interesting approach, I am not sure I understand what is going on exactly.

Does your Lua script simply replace the ModSecurity body processor to keep the  `REQUEST_BODY` target populated while still extracting the elements from the XML body?

Are the `SecRuleUpdateTargetById` directives necessary because your Lua-based body processor populates a different collection?

Does your custom xml collection allow addressing elements using XPath expressions or are they addressed differently?

**Comment by azurit** [comment]
@studersi

> Does your Lua script simply replace the ModSecurity body processor to keep the `REQUEST_BODY` target populated while still extracting the elements from the XML body?

Yes, exactly.

> Are the `SecRuleUpdateTargetById` directives necessary because your Lua-based body processor populates a different collection?

Yes. Unfortunately i cannot create or edit original `XML` collection so i populated `tx.xml` (but it can be anything under `tx.*`).

> Does your custom xml collection allow addressing elements using XPath expressions or are they addressed differently?

Sorry, i'm not expert on XML. If you mean 'addressing' like `TX:xml.userInfo.lastName` then yes, see last error message:
`Matched Data: union select found within TX:xml.userInfo.lastName: union select`

**Comment by studersi** [comment]
I see, that clears things up a bit, thanks.

The [`XML`](https://github.com/SpiderLabs/ModSecurity/wiki/Reference-Manual-%28v2.x%29#xml) collection allow you to select elements using XPath expressions which is a powerful query language for XML.

Would it be possible to store the content of the `REQUEST_BODY` collection in a transient collection (e.g. `tx.request_body`), use this transient variable for the XXE rules, and then run the regular XML body processor instead?

**Comment by azurit** [comment]
> Would it be possible to store the content of the `REQUEST_BODY` collection in a transient collection (e.g. `tx.request_body`), use this transient variable for the XXE rules, and then run the regular XML body processor instead?

Unfortunately no because:
 - any body processor MUST be activated in phase 1, cannot be activated later
 - REQUEST_BODY variable is accessible from phase 2 (even within Lua)

**Comment by azurit** [comment]
Now i see the problem with XPath expressions you were asking about: No, it's not working and probably cannot be done (at least not in backward compatible way). But you can use regular expressions (not the same but may be sufficient):
`TX:/^xml\.any-regex-here/`

**Comment by studersi** [comment]
It might actually not be a problem. I glanced over the usages of the XML collection and the most complex XPath expressions I could spot were: `...|XML:/*|XML://@*`. I think these expressions select the values of all elements and all attributes of all elements. This could be feasible with a simpler selector like regexes.

**Comment by dune73** [comment]
Please elaborate on the "This could be feasible with a simpler selector like regexes."

**Comment by studersi** [comment]
Simple XPath expressions like the ones used in CRS for the XML collection could be implemented using Regexes as far as I can tell as long as the elements and their attributes are available in the TX collection. An example for this was given by @azurit.

This would not necessarily have been the case if CRS made use of more complex XPath expression.

**Comment by dune73** [comment]
Ah, got you. So as long as we target every element, we can select them via regex selection on `TX:/.../`? Neat thought.

**Comment by studersi** [comment]
Exactly. As long as we select all elements and attributes, it should be easy to translate the XPath expressions into corresponding regexes.Given how generic CRS is this should cover all cases.

Problems could arise where there are application-specific ModSecurity rules outside of CRS that rely on this functionality or the XML collection in general.

**Comment by azurit** [comment]
Should i continue in development of this plugin? I mean, will it be usable for us in current form?

**Comment by dune73** [comment]
I see some real potential yes. So if you are interested, then please do.

@studersi : What is your take? Any chance your customers would run this midterm?

**Comment by azurit** [comment]
> On top: Not sure how FP handling integrates into this approach.

Exclusion rules written for `XML` collection won't be compatible.

**Comment by studersi** [comment]
So far, I have never seen anyone use XPath expressions to create custom XML rules tailored for individual payloads. I have also not seen anyone use the XML schema validation features that would require the XML body processor.

Switching the rule exclusions from the XML collection to the new TX:xml collections would be doable.

It is hard to say whether the plugin would be attractive for our customers.

There is obviously somewhat of a risk in switching to the plugin for XML handling as it is new and replaces a fairly important and complex part of the ModSecurity engine and XXEs should be trivial to disable application-wide on the backend if not in use.

On the other hand, XXEs can be quite catastrophic if not mitigated properly or if the devs forget to turn XXEs off.

**Comment by dune73** [comment]
Thank you for this balanced view.

One option could be to offer both body processors and enabling one of them according to a local policy and information about the backend.

**Comment by azurit** [comment]
Attaching new version of XXE plugin, this one is a big step towards backward compatibility:
[xxe.tar.gz](https://github.com/coreruleset/coreruleset/files/8573240/xxe.tar.gz)

New version supports access to:
 - values from XML elements as a string via `TX:XML`
 - XML element attributes as a collection via `TX:XMLATTRS`

Both of them should be in the same form as using original XML parser but accessible under different name (which makes rule exclusions not compatible by the way).

Example of output from original XML parser (first) vs. new while parsing simple XML document:

```
ModSecurity: Warning. Unconditional match in SecAction. [file "/usr/share/modsecurity-crs-new/plugins/dump-after.conf"] [line "7"] [id "100"] [data "Matched Data: \\x0a    \\x0a        Fred Jones\\x0a        \\x0a            900 Aurora Ave.\\x0a            Seattle\\x0a            WA\\x0a            98115\\x0a        \\x0a        \\x0a            2011 152nd Avenue NE\\x0a            Redmond\\x0a            WA\\x0a            98052\\x0a        \\x0a        (425)555-5665\\x0a        (206)555-5555\\x0a        (206)555-4321\\x0a    \\x0a found within XML"] [hostname "example.com"] [uri "/"] [unique_id "YmlZ3UvMjQfUf_QBeOThewAAAGs"]
ModSecurity: Warning. Unconditional match in SecAction. [file "/usr/share/modsecurity-crs-new/plugins/dump-after.conf"] [line "16"] [id "101"] [data "Matched Data: home found within XML"] [hostname "example.com"] [uri "/"] [unique_id "YmlZ3UvMjQfUf_QBeOThewAAAGs"]
ModSecurity: Warning. Unconditional match in SecAction. [file "/usr/share/modsecurity-crs-new/plugins/dump-after.conf"] [line "16"] [id "101"] [data "Matched Data: true found within XML"] [hostname "example.com"] [uri "/"] [unique_id "YmlZ3UvMjQfUf_QBeOThewAAAGs"]
ModSecurity: Warning. Unconditional match in SecAction. [file "/usr/share/modsecurity-crs-new/plugins/dump-after.conf"] [line "16"] [id "101"] [data "Matched Data: work found within XML"] [hostname "example.com"] [uri "/"] [unique_id "YmlZ3UvMjQfUf_QBeOThewAAAGs"]
ModSecurity: Warning. Unconditional match in SecAction. [file "/usr/share/modsecurity-crs-new/plugins/dump-after.conf"] [line "16"] [id "101"] [data "Matched Data: work found within XML"] [hostname "example.com"] [uri "/"] [unique_id "YmlZ3UvMjQfUf_QBeOThewAAAGs"]
ModSecurity: Warning. Unconditional match in SecAction. [file "/usr/share/modsecurity-crs-new/plugins/dump-after.conf"] [line "16"] [id "101"] [data "Matched Data: home found within XML"] [hostname "example.com"] [uri "/"] [unique_id "YmlZ3UvMjQfUf_QBeOThewAAAGs"]
ModSecurity: Warning. Unconditional match in SecAction. [file "/usr/share/modsecurity-crs-new/plugins/dump-after.conf"] [line "16"] [id "101"] [data "Matched Data: mobile found within XML"] [hostname "example.com"] [uri "/"] [unique_id "YmlZ3UvMjQfUf_QBeOThewAAAGs"]

```
```
ModSecurity: Warning. Unconditional match in SecAction. [file "/usr/share/modsecurity-crs-new/plugins/dump-after.conf"] [line "34"] [id "105"] [data "Matched Data: \\x0a    \\x0a        Fred Jones\\x0a        \\x0a            900 Aurora Ave.\\x0a            Seattle\\x0a            WA\\x0a            98115\\x0a        \\x0a        \\x0a            2011 152nd Avenue NE\\x0a            Redmond\\x0a            WA\\x0a            98052\\x0a        \\x0a        (425)555-5665\\x0a        (206)555-5555\\x0a        (206)555-4321\\x0a    \\x0a found within TX:xml"] [hostname "example.com"] [uri "/"] [unique_id "YmlZ1CTFhRT7QXGwOSIudQAAAGY"]
ModSecurity: Warning. Unconditional match in SecAction. [file "/usr/share/modsecurity-crs-new/plugins/dump-after.conf"] [line "43"] [id "106"] [data "Matched Data: home found within TX:xmlattrs.0"] [hostname "example.com"] [uri "/"] [unique_id "YmlZ1CTFhRT7QXGwOSIudQAAAGY"]
ModSecurity: Warning. Unconditional match in SecAction. [file "/usr/share/modsecurity-crs-new/plugins/dump-after.conf"] [line "43"] [id "106"] [data "Matched Data: true found within TX:xmlattrs.1"] [hostname "example.com"] [uri "/"] [unique_id "YmlZ1CTFhRT7QXGwOSIudQAAAGY"]
ModSecurity: Warning. Unconditional match in SecAction. [file "/usr/share/modsecurity-crs-new/plugins/dump-after.conf"] [line "43"] [id "106"] [data "Matched Data: work found within TX:xmlattrs.2"] [hostname "example.com"] [uri "/"] [unique_id "YmlZ1CTFhRT7QXGwOSIudQAAAGY"]
ModSecurity: Warning. Unconditional match in SecAction. [file "/usr/share/modsecurity-crs-new/plugins/dump-after.conf"] [line "43"] [id "106"] [data "Matched Data: work found within TX:xmlattrs.3"] [hostname "example.com"] [uri "/"] [unique_id "YmlZ1CTFhRT7QXGwOSIudQAAAGY"]
ModSecurity: Warning. Unconditional match in SecAction. [file "/usr/share/modsecurity-crs-new/plugins/dump-after.conf"] [line "43"] [id "106"] [data "Matched Data: home found within TX:xmlattrs.4"] [hostname "example.com"] [uri "/"] [unique_id "YmlZ1CTFhRT7QXGwOSIudQAAAGY"]
ModSecurity: Warning. Unconditional match in SecAction. [file "/usr/share/modsecurity-crs-new/plugins/dump-after.conf"] [line "43"] [id "106"] [data "Matched Data: mobile found within TX:xmlattrs.5"] [hostname "example.com"] [uri "/"] [unique_id "YmlZ1CTFhRT7QXGwOSIudQAAAGY"]

```

Side note: Are we using XPath searching correctly? We search only for two patterns:
1. `XML:/*`: This is outputting all values from XML elements as a one big string, usually delimited by whitespaces and/or newlines.
2. `XML://@*`: This is outputting a collection/array of attribute values.

Don't understand why one is a string and other a collection.

**Comment by studersi** [comment]
@azurit I see, so now the attributes are covered as well.

> `XML:/*`: This is outputting all values from XML elements as a one big string, usually delimited by whitespaces and/or newlines.

I have wondered this as well and would assume that `XML://*` would have been more appropriate. However, since in my tests the rules also matched for nested payloads so I did not investigate any further. I was not aware that there is such a fundamental difference. I am wondering now if there could be performance difference matching everything as a single string compared to separate targets.

Also, do you know if there are any pertinent differences between the _expat_ XML library instead of _libxml2_ that the standard body processor uses?

**Comment by azurit** [comment]
> Also, do you know if there are any pertinent differences between the _expat_ XML library instead of _libxml2_ that the standard body processor uses?

@studersi Unfortunately no. More complex XML documents needs to be tested and output compared to see if there are any differences between parsers outputs.

**Comment by azurit** [comment]
@studersi I found [XMLua](https://github.com/clear-code/xmlua) which is based on libxml2 and looks very good. Would be probably, from the point of compatibility, better to use XML parser based on the same C library as original one. Unfortunately, it looks like it's not part of any linux distribution (can be installed using luarocks though). Not sure which one to use, what do you think?

**Comment by studersi** [comment]
Ah, that would complicate the installation process considerably I would assume....

A random thought just came up. Could we leverage [STREAM_INPUT_BODY](https://github.com/SpiderLabs/ModSecurity/wiki/Reference-Manual-%28v2.x%29#stream_input_body) for XXE rules?
> This variable give access to the raw request body content. This variable is best used for two use-cases:
> 1. For fast pattern matching - using @pm/@pmf to prequalify large text strings against any kind of content-type data. This is more performant vs. using REQUEST_BODY/ARGS_POST/ARGS_POST_NAMES as it happens before ModSecurity parsing in phase:2 variable population.
> 2. For data substitution - using @rsub against this variable allows you to manipulate live request body data. Example - to remove offending payloads or to substitute benign data.

**Comment by azurit** [comment]
Well, this actually works! Attaching alternative version of XXE plugin:
[xxe2-before.conf.gz](https://github.com/coreruleset/coreruleset/files/8585225/xxe2-before.conf.gz)

Few notes:
 - no Lua, pure SecLang
 - for ModSecurity v2, setting `SecStreamInBodyInspection` needs to be set to `On` with this warning from docs:
` This directive may significantly impact file upload times. The impact depends on server resources and the nature of operations being performed on the request bodies being streamed in.`
 - for ModSecurity v3, `SecStreamInBodyInspection`/`STREAM_INPUT_BODY` isn't supported but thanks to a accidental feature, `REQUEST_BODY` is not wiped while using body parsers, so i included it in targets for v3 compatibility
 
None of the plugins is perfect but it depends on the use-case.

**Comment by studersi** [comment]
Thanks a lot, I will try do some tests with this to see what we can expect in terms of performance impact.

For applications where the performance impact is acceptable, I would assume that this version is much more palatable than replacing the entire body processor.

**Comment by dune73** [comment]
Wow, if this works, then we have a super-dirty but working solution as long as the v3 behavior remains inconsistent with v2.

**Comment by azurit** [comment]
Attaching new versions of both plugins:
[xxe.tar.gz](https://github.com/coreruleset/coreruleset/files/8595604/xxe.tar.gz)
[xxe2-before.conf.gz](https://github.com/coreruleset/coreruleset/files/8595605/xxe2-before.conf.gz)

Changes for both:
- support for two new entity declarations:
```
<!ENTITY name PUBLIC "public_ID" "URI">
<!ENTITY % name SYSTEM "URI">
```
- new rule trying to catch DoS attempts
- more enhancement based on these [payloads](https://github.com/payloadbox/xxe-injection-payload-list)

Changes for lua version of the plugin:
- do not run XML parser if any payloads are found in `REQUEST_BODY`

I must say that i like lua version more. For example, imagine a security bug in libxml2 which can be triggered only by parsing the document. Abusing it you can attack modsecurity directly without any option of defense. But using a custom XML parser you have more control over when to enable it - we really shouldn't parse an XML document if we know it contains payloads.

**Comment by azurit** [comment]
Related to slow uploads while using `SecStreamInBodyInspection`:
https://github.com/SpiderLabs/ModSecurity/issues/1366

**Comment by azurit** [comment]
Any news here?

**Comment by studersi** [comment]
@azurit That is indeed quite interesting!

It would probably be best to not run the XML parser if we already know there are XML-based attacks present. [CVE-2013-1915](https://nvd.nist.gov/vuln/detail/CVE-2013-1915) spontaneously comes to mind.

As far as I can tell it would be entirely novel that CRS protects the engine it runs on. This might be something worth exploring. The only downside I see with this is that we would almost have to block the request immediately in such a case. Not running the XML parser could end with a bypass of the rules that look at the parsed XML targets. What do you think @dune73?

As for the alternative implementation, it might make sense to maintain two versions of this plugin, one with the custom XML parsing and one without. I would assume there is a market for both of them...

**Comment by azurit** [comment]
I'm able to maintain both versions.

**Comment by dune73** [comment]
I agree on the interest in two separate plugins for this. It's also a bit of an experimental field and offering two options would allow us to feel where the interest is. So yes, it's a cool test.

**Comment by dune73** [comment]
New idea. Ideal to explore it in a plugin and see if it works. The blocking makes sense in this situation and there is nothing that prohibits a plugin from blocking directly if there is a real reason.

**Comment by azurit** [comment]
Here are both plugins, fully complete, including documentation:
https://github.com/azurit/modsecurity-xxe-plugin
https://github.com/azurit/modsecurity-xxe-lua-plugin

I'm not sure about the names of plugins, what do you think?

@dune73 if everything is ok i would ask for two new repositories, thank you.

**Comment by dune73** [comment]
This looks good.

I think you need to indicate the engine you are aiming for in the title. Otherwise it's too puzzling for people who are not familiar with the topic. I'd do "ModSecurity-v2-xxe-plugin" and "ModSecurity-v3-xxe-lua-plugin". Or maybe "ModSecurity2-..." etc.

Happy to create the repos after the renaming.

**Comment by azurit** [comment]
The lua version works on both, v2 and v3.

**Comment by dune73** [comment]
Oh, nice twist.

@studersi what do you think about the naming here?

**Comment by azurit** [comment]
Maybe:
xxe-ModSec2
xxe-lua

**Comment by studersi** [comment]
> Maybe:
> xxe-ModSec2
> xxe-lua

Why ModSec2 if they work for both versions?

**Comment by azurit** [comment]
[This one](https://github.com/azurit/modsecurity-xxe-plugin) works only in ModSec v2.

**Comment by studersi** [comment]
Oh, I misread that they both work on both engines. In that case, I guess that naming would be appropriate.

**Comment by azurit** [comment]
I have one issue with Lua version of the plugin.

To make it all work and keep backward compatible, plugin is using `SecRuleRemoveById` and `SecRuleUpdateTargetById` to modify current rules on the fly (see [here](https://github.com/azurit/modsecurity-xxe-lua-plugin/blob/main/plugins/xxe-lua-after.conf)). Problem is that these directives cannot be disabled using SecRule language so, if the plugin is installed using [Method 1](https://coreruleset.org/docs/concepts/plugins/#method-1-copying-the-plugin-files), part of it (for example removing of rule `200000`) will remain enabled even if plugin is disabled.

Is it sufficient to only mention this in the docs and recommend [Method 2](https://coreruleset.org/docs/concepts/plugins/#method-2-placing-symbolic-links-to-separate-plugin-files-downloaded-elsewhere) of plugin installation?

**Comment by azurit** [comment]
ping

**Comment by theseion** [comment]
> Is it sufficient to only mention this in the docs and recommend [Method 2](https://coreruleset.org/docs/concepts/plugins/#method-2-placing-symbolic-links-to-separate-plugin-files-downloaded-elsewhere) of plugin installation?

I think yes. Installing plugins requires cloning / fetching from the plugin repositories explicitly. I think we can expect users to pay enough attention to read the documentation of the plugins they want to install.

Greaty work BTW! 🚀

**Comment by fzipi** [comment]
@azurit two things, following up on my close request in July.

First, on what is left here. From the thread, both plugins were finished in August 2022, @dune73 approved them, the naming was settled as `xxe-ModSec2` / `xxe-lua`, and @theseion answered the last open question (the `SecRuleRemoveById` / installation method caveat) in 2024. The only step that never happened is creating the org repositories. If that is all that remains, say so and I will get the repos created and the code transferred. If something else is still blocking, it would help to have it written down — the thread does not say.

Second, a data point that changes one of the premises here. I measured the entity declaration across all three engines in #4767:

| | v2 / Apache | v3 / nginx | Coraza |
| --- | --- | --- | --- |
| `REQUEST_BODY` has `<!ENTITY` | no | **yes** | no (empty) |

On libmodsecurity3 the declaration is in `REQUEST_BODY` and `FULL_REQUEST`, and a plain `SecRule REQUEST_BODY "@rx (?i)<!\s*(?:doctype|entity)"` catches both the classic and the blind parameter-entity shape with no Lua involved. This is the v2/v3 discrepancy you and @studersi discussed here in 2022 — it is still there.

That does not remove the need for the v2 plugin, and it does not make the Lua plugin pointless (it still buys the full `XML` collection). But it does mean the v3 case may not need a plugin at all, which is worth factoring into how these are documented and scoped. Details and the full engine table are in #4767.

---

- **tools used**: Claude (Opus 5), via Claude Code
- **assisted with**: re-reading this thread for the outstanding blocker, running the cross-engine measurement, drafting this comment
- **review performed**: measured the variable state on live `modsec2-apache`, `modsec3-nginx` and `coraza-caddy` from `tests/docker-compose.yml` on CRS `main` with purpose-written phase 2 rules, and confirmed the candidate `REQUEST_BODY` rule matches on v3 and not on v2 for both XXE shapes; re-read the full comment history here to identify which steps were completed and which were not

**Comment by azurit** [comment]
@fzipi There are actually two plugins:
 - [one which uses Lua](https://github.com/azurit/modsecurity-xxe-lua-plugin)
 - [one which does not use Lua but needs SecStreamInBodyInspection=On](https://github.com/azurit/modsecurity-xxe-plugin)

**Comment by fzipi** [comment]
I feel like we still need to document this better at least. And I'll try to push for a fix on the coraza side.
